File: src/utility/Shell.ts
/**
 * Shell Utility - Execute system commands with proper error handling
 *
 * Features:
 * - Async-only API for non-blocking operations
 * - Timeout support with graceful termination
 * - Structured result with exit code, stdout, stderr
 * - Optional streaming callbacks for real-time output
 * - Integration with Logger for command logging
 */

import { spawn } from "node:child_process";
import { Logger } from "./Logger";

export interface ShellResult {
	exitCode: number;
	stdout: string;
	stderr: string;
	timedOut: boolean;
	command: string;
}

export interface ShellOptions {
	timeout?: number; // Timeout in milliseconds (default: 30000)
	onStdout?: (line: string) => void; // Callback for each stdout line
	onStderr?: (line: string) => void; // Callback for each stderr line
	cwd?: string; // Working directory
	env?: Record<string, string>; // Environment variables
	logCommand?: boolean; // Log command execution (default: true)
}

export class Shell {
	private static logger = Logger.getInstance();

	private constructor() {
		// Prevent instantiation - this is a static utility class
	}

	/**
	 * Execute a shell command asynchronously
	 *
	 * @param command - The command to execute (passed to bash -c)
	 * @param options - Execution options
	 * @returns Promise resolving to ShellResult
	 */
	static async execute(
		command: string,
		options: ShellOptions = {},
	): Promise<ShellResult> {
		const {
			timeout = 30000,
			onStdout,
			onStderr,
			cwd,
			env,
			logCommand = true,
		} = options;

		if (logCommand) {
			this.logger.debug(`Executing: ${command}`);
		}

		return new Promise((resolve) => {
			const proc = spawn("bash", ["-c", command], {
				stdio: ["ignore", "pipe", "pipe"],
				cwd,
				env: { ...process.env, ...env },
			});

			let stdout = "";
			let stderr = "";
			let completed = false;
			let timeoutHandle: NodeJS.Timeout | undefined;

			// Setup timeout handler
			if (timeout > 0) {
				timeoutHandle = setTimeout(() => {
					if (!completed) {
						completed = true;
						proc.kill("SIGTERM");
						this.logger.warn(
							`Command timed out after ${timeout}ms: ${command}`,
						);
						resolve({
							exitCode: -1,
							stdout,
							stderr: stderr || `Timed out after ${timeout}ms`,
							timedOut: true,
							command,
						});
					}
				}, timeout);
			}

			// Capture stdout
			proc.stdout.on("data", (data: Buffer) => {
				const text = data.toString();
				stdout += text;
				if (onStdout) {
					text.split("\n").forEach((line) => {
						if (line.trim()) onStdout(line);
					});
				}
			});

			// Capture stderr
			proc.stderr.on("data", (data: Buffer) => {
				const text = data.toString();
				stderr += text;
				if (onStderr) {
					text.split("\n").forEach((line) => {
						if (line.trim()) onStderr(line);
					});
				}
			});

			// Handle process completion
			proc.on("close", (code) => {
				if (!completed) {
					completed = true;
					if (timeoutHandle) clearTimeout(timeoutHandle);
					resolve({
						exitCode: code || 0,
						stdout,
						stderr,
						timedOut: false,
						command,
					});
				}
			});

			// Handle process errors
			proc.on("error", (err) => {
				if (!completed) {
					completed = true;
					if (timeoutHandle) clearTimeout(timeoutHandle);
					this.logger.error(`Command error: ${err.message}`);
					resolve({
						exitCode: -1,
						stdout,
						stderr: err.message,
						timedOut: false,
						command,
					});
				}
			});
		});
	}
}

File: src/utility/Shell.go
package utility

import (
	"bufio"
	"bytes"
	"context"
	"fmt"
	"os/exec"
	"strings"
	"time"
)

// Shell provides command execution capabilities
type Shell struct {
	logger *Logger
}

// Result contains the output of a command execution
type Result struct {
	ExitCode int
	Stdout   string
	Stderr   string
	TimedOut bool
	Duration time.Duration
	Command  string
}

// ExecOptions configures command execution
type ExecOptions struct {
	Timeout        time.Duration
	StdoutCallback func(line string)
	StderrCallback func(line string)
	Env            map[string]string
	WorkDir        string
	UseSudo        bool
}

// NewShell creates a new Shell executor
func NewShell(logger *Logger) *Shell {
	return &Shell{logger: logger}
}

// Execute runs a command with the given options
func (s *Shell) Execute(ctx context.Context, command string, opts *ExecOptions) (*Result, error) {
	if opts == nil {
		opts = &ExecOptions{
			Timeout: 30 * time.Second,
		}
	}

	// Set default timeout if not specified
	if opts.Timeout == 0 {
		opts.Timeout = 30 * time.Second
	}

	// Add sudo if requested
	if opts.UseSudo {
		command = fmt.Sprintf("sudo %s", command)
	}

	// Create context with timeout
	execCtx := ctx
	if opts.Timeout > 0 {
		var cancel context.CancelFunc
		execCtx, cancel = context.WithTimeout(ctx, opts.Timeout)
		defer cancel()
	}

	startTime := time.Now()

	// Create command
	cmd := exec.CommandContext(execCtx, "bash", "-c", command)

	// Set working directory
	if opts.WorkDir != "" {
		cmd.Dir = opts.WorkDir
	}

	// Set environment variables
	if len(opts.Env) > 0 {
		cmd.Env = append(cmd.Env, s.envMapToSlice(opts.Env)...)
	}

	// Create stdout and stderr pipes
	stdoutPipe, err := cmd.StdoutPipe()
	if err != nil {
		return nil, fmt.Errorf("failed to create stdout pipe: %w", err)
	}

	stderrPipe, err := cmd.StderrPipe()
	if err != nil {
		return nil, fmt.Errorf("failed to create stderr pipe: %w", err)
	}

	// Start the command
	if err := cmd.Start(); err != nil {
		return nil, fmt.Errorf("failed to start command: %w", err)
	}

	// Capture stdout
	var stdoutBuf bytes.Buffer
	stdoutDone := make(chan struct{})
	go func() {
		scanner := bufio.NewScanner(stdoutPipe)
		for scanner.Scan() {
			line := scanner.Text()
			stdoutBuf.WriteString(line + "\n")
			if opts.StdoutCallback != nil {
				opts.StdoutCallback(line)
			}
		}
		close(stdoutDone)
	}()

	// Capture stderr
	var stderrBuf bytes.Buffer
	stderrDone := make(chan struct{})
	go func() {
		scanner := bufio.NewScanner(stderrPipe)
		for scanner.Scan() {
			line := scanner.Text()
			stderrBuf.WriteString(line + "\n")
			if opts.StderrCallback != nil {
				opts.StderrCallback(line)
			}
		}
		close(stderrDone)
	}()

	// Wait for output reading to complete
	<-stdoutDone
	<-stderrDone

	// Wait for command to complete
	err = cmd.Wait()
	duration := time.Since(startTime)

	result := &Result{
		ExitCode: 0,
		Stdout:   strings.TrimSpace(stdoutBuf.String()),
		Stderr:   strings.TrimSpace(stderrBuf.String()),
		TimedOut: false,
		Duration: duration,
		Command:  command,
	}

	// Check if command timed out
	if execCtx.Err() == context.DeadlineExceeded {
		result.TimedOut = true
		result.ExitCode = -1
		return result, fmt.Errorf("command timed out after %v", opts.Timeout)
	}

	// Get exit code
	if err != nil {
		if exitErr, ok := err.(*exec.ExitError); ok {
			result.ExitCode = exitErr.ExitCode()
		} else {
			result.ExitCode = -1
			return result, fmt.Errorf("command failed: %w", err)
		}
	}

	return result, nil
}

// envMapToSlice converts a map of environment variables to a slice
func (s *Shell) envMapToSlice(env map[string]string) []string {
	result := make([]string, 0, len(env))
	for key, value := range env {
		result = append(result, fmt.Sprintf("%s=%s", key, value))
	}
	return result
}

// QuickExec is a convenience method for simple command execution
func (s *Shell) QuickExec(command string) (*Result, error) {
	return s.Execute(context.Background(), command, nil)
}

// ExecWithTimeout runs a command with a specific timeout
func (s *Shell) ExecWithTimeout(command string, timeout time.Duration) (*Result, error) {
	return s.Execute(context.Background(), command, &ExecOptions{Timeout: timeout})
}

// ExecWithSudo runs a command with sudo
func (s *Shell) ExecWithSudo(command string) (*Result, error) {
	return s.Execute(context.Background(), command, &ExecOptions{UseSudo: true})
}
